sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp

```diff
@@ -0,0 +1,118 @@
+    std::unique_ptr<Azure::Core::Http::RawResponse> m_rawResponse;
+    std::string m_continuationToken;
+
+    /* This is the implementation for checking the status of a deleted key. The key is considered
+     * deleted if querying /deletedkeys/keyName returns 200 from server. Or whenever soft-delete is
+     * disabled.*/
+    std::unique_ptr<Azure::Core::Http::RawResponse> PollInternal(
+        Azure::Core::Context& context) override;
+
+    Azure::Core::Response<Azure::Security::KeyVault::Keys::DeletedKey> PollUntilDoneInternal(
+        Azure::Core::Context& context,
+        std::chrono::milliseconds period) override
+    {
+      while (!IsDone())
+      {
+        m_rawResponse = Poll(context);
+        std::this_thread::sleep_for(period);
+      }
+
+      return Azure::Core::Response<Azure::Security::KeyVault::Keys::DeletedKey>(
+          m_value, std::make_unique<Azure::Core::Http::RawResponse>(*m_rawResponse));
+    }
+
+    /*
+     * Only friend classes are permitted to construct a DeleteOperation. This is because a
+     * KeyVaultPipelne is required and it is not exposed to customers.
+     *
+     * Since C++ doesn't offer `internal` access, we use friends-only instead.
+     */
+    DeleteKeyOperation(
+        std::shared_ptr<Azure::Security::KeyVault::Common::Internal::KeyVaultPipeline>
+            keyvaultPipeline,
+        Azure::Core::Response<Azure::Security::KeyVault::Keys::DeletedKey> response)
+        : m_pipeline(keyvaultPipeline)
+    {
+      if (!response.HasValue())
+      {
+        throw Azure::Security::KeyVault::Common::KeyVaultException(
+            "The response does not contain a value.");
+      }
+      // The response becomes useless and the value and rawResponse are now owned by the
+      // DeleteKeyOperation. This is fine because the DeleteKeyOperation is what the delete key api
+      // will return.
+      m_value = response.ExtractValue();
+      m_rawResponse = response.ExtractRawResponse();
+
+      // Build the full url for continuation token. It is only used in case customers wants to use
+      // it on their own. The Operation uses the KeyVaultPipeline from the client which knows how to
+      // build this url.
+      m_continuationToken = m_pipeline->GetVaultUrl() + "/" + std::string(Details::DeletedKeysPath)
+          + "/" + m_value.Name();
+
+      // The recoveryId is only returned if soft-delete is enabled.
+      // The LRO is considered completed for non soft-delete (key will be eventually removed).
+      if (m_value.RecoveryId.empty())
+      {
+        m_status = Azure::Core::OperationStatus::Succeeded;
+      }
+    }
+
+  public:
+    /**
+     * @brief Get the #DeletedKey object.
+     *
+     * @remark The deleted key contains the recovery id if the key can be recovered.
+     *
+     * @return A deleted key object.
+     */
+    Azure::Security::KeyVault::Keys::DeletedKey Value() const override { return m_value; }
```

Comment:
It may be more idiomatic to do it this way, but with .NET we actually try to avoid throwing exceptions from the update handler (your poller above) and do so here, but that's mimicking nullable types in .NET more. I just wanted to point out the difference. May not matter here.

Reply:
We throw in constructor if values is received without a value

```cpp
if (!response.HasValue())
      {
        throw Azure::Security::KeyVault::Common::KeyVaultException(
            "The response does not contain a value.");
      }
```

The polling update (we call it just Poll() ) would throw only if the operation is cancelled.